### PR TITLE
[IMP] account, l10n_*: Introduce unified reporting engine

### PR DESCRIPTION
**Pull request by oco-odoo**
[IMP] account, l10n_*: Introduce unified reporting engine

This commit adapts account's model to the new report engine introduced for v16, and updates the data files accordingly. 

account.report model is now declared in community, together with the other models used by the reporting. This is done so that the tax tags can properly be created by the tax report and used on tax templates. All the actual computation logic stays in enterprise. 

See enterprise commit for full details.

Task 2524389

### CODE REVIEW

File: addons/account/models/chart_template.py [hunk 1]
```diff
@@ -1215,36 +1215,33 @@ def create(self, vals_list):
     def _convert_tag_syntax_to_orm(self, tags_list):
         """ Repartition lines give the possibility to directly give
         a list of ids to create for tags instead of a list of ORM commands.
-
```
Comment (smetl) [inline addons/account/models/chart_template.py:1218]: useless diff

File: addons/account/models/chart_template.py [hunk 2]
```diff
@@ -1215,36 +1215,33 @@ def create(self, vals_list):
...
         This function checks that tags_list uses this syntactic sugar and returns
         an ORM-compliant version of it if it does.
         """
         if tags_list and all(isinstance(elem, int) for elem in tags_list):
             return [(6, False, tags_list)]
-        return tags_list
+        return
```
Comment (smetl) [inline addons/account/models/chart_template.py:1223]: In that case, you are setting `None` for a m2m. I'm not sure the ORM is not crying in that case..

File: addons/account/security/ir.model.access.csv [hunk 3]
```diff
@@ -2,7 +2,7 @@ id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
 access_account_cash_rounding_readonly,account.cash.rounding,model_account_cash_rounding,account.group_account_readonly,1,0,0,0
 access_account_cash_rounding_uinvoice,account.cash.rounding,model_account_cash_rounding,account.group_account_invoice,1,1,1,1
 access_res_partner_group_account_manager,res_partner group_account_manager,model_res_partner,account.group_account_manager,1,0,0,0
-
+,,,,,,,
```
Comment (smetl) [inline addons/account/security/ir.model.access.csv:5]: wtf

File: addons/web/static/src/webclient/actions/action_service.js [hunk 4]
```diff
@@ -753,7 +753,8 @@ function makeActionManager(env) {
             componentProps: controller.props,
         };
         env.bus.trigger("ACTION_MANAGER:UPDATE", controller.__info__);
-        return Promise.all([currentActionProm, closingProm]).then((r) => r[0]);
+        let res = Promise.all([currentActionProm, closingProm]).then((r) => r[0]);
+        return res;
```
Comment (smetl) [inline addons/web/static/src/webclient/actions/action_service.js:757]: remove this diff

File: odoo/models.py [hunk 5]
```diff
@@ -4868,6 +4868,7 @@ def _search(self, domain, offset=0, limit=None, order=None, count=False, access_
 
         if expression.is_false(self, domain):
             # optimization: no need to query, as no record satisfies the domain
+
```
[the remark this answers is not in the document]
Comment (smetl) [reply odoo/models.py:4871]: useless diff

**Review by smetl — COMMENTED** [review]
(note: I didn't review the l10n* part)

File: addons/account/models/account_report.py [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+
+class AccountReportExpression(models.Model):
+    _name = "account.report.expression"
+    _description = "Accounting Report Expression"
+
+    report_line_id = fields.Many2one(string="Report Line", comodel_name='account.report.line', required=True, ondelete='cascade')
+    label = fields.Char(string="Label", required=True)
+    engine = fields.Selection(
+        string="Computation Engine",
+        selection = [
```
Comment (smetl) [inline addons/account/models/account_report.py:330]: remove extra whitespaces

File: addons/account/models/account_report.py [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+    report_country_id = fields.Many2one(string="Country", related='target_report_line_id.report_id.country_id')
+
+    company_id = fields.Many2one(string='Company', comodel_name='res.company', required=True, default=lambda self: self.env.company)
+
+    foreign_vat_fiscal_position_id = fields.Many2one(
+        string="Fiscal position",
+        comodel_name='account.fiscal.position',
+        domain="[('company_id', '=', company_id), ('country_id', '=', report_country_id), ('foreign_vat', '!=', False)]",
+        help="The foreign fiscal position for which this external value is made.",
+    )
```
Comment (smetl) [inline addons/account/models/account_report.py:549]: need a `check_company`

File: addons/account/models/account_report.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+
+    _sql_constraints = [
+        ('code_uniq', 'unique (code)', "A report line with the same code already exists."),
+    ]
+
+    @api.depends('parent_id.hierarchy_level')
+    def _compute_hierarchy_level(self):
+        for record in self:
+            if record.parent_id:
+                record.hierarchy_level = record.parent_id.hierarchy_level + 2
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/models/account_report.py:234]: Yup, that's what we used to do for the tax report ;)

File: addons/account/models/account_report.py [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+    #  CORE ==========================================================================================================================================
+
+    name = fields.Char(string="Name", required=True, translate=True)
+    line_ids = fields.One2many(string="Lines", comodel_name='account.report.line', inverse_name='report_id')
+    column_ids = fields.One2many(string="Columns", comodel_name='account.report.column', inverse_name='report_id')
+    root_report_id = fields.Many2one(string="Root Report", comodel_name='account.report')
+    variant_report_ids = fields.One2many(string="Variants", comodel_name='account.report', inverse_name='root_report_id')
+    country_id = fields.Many2one(string="Country", comodel_name='res.country')
+    only_tax_exigible = fields.Boolean(string="Only Tax Exigible Lines", default=False, required=True)
+    caret_options_initializer = fields.Char(string="Caret Options Initializer", required=True, default='_get_default_caret_options')
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/models/account_report.py:35]: It's a Char field, so no warning should be raised. The function is in enterprise, like all the content of the engine, it's not a problem. 

All reports have a caret options *initializer*, which does not mean all of them have caret options. The caret options initializer is used to map the caret optinos label to actual options.

File: addons/account/models/account_report.py [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+        string="Carry Over To",
+        help="Formula in the form line_code.expression_label. This allows setting the target of the carryover for this expression "
+             "(on a _carryover_*-labeled expression), in case it is different from the parent line. "
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        rslt = super(AccountReportExpression, self).create(vals_list)
+
+        for record in rslt:
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/models/account_report.py:372]: Agree to disagree :p

File: addons/account/models/account_report.py [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+        if 'country_id' in vals:
+            tags_cache = {}
+            impacted_reports = self.filtered(lambda x: x.country_id.id != vals['country_id'])
+            tax_tags_expressions = impacted_reports.line_ids.expression_ids.filtered(lambda x: x.engine == 'tax_tags')
+
+            for expression in tax_tags_expressions:
+                tax_tags = self.env['account.account.tag']._get_tax_tags(expression.formula, expression.report_line_id.report_id.country_id.id)
+                tag_reports = tax_tags._get_related_tax_report_expressions().report_line_id.report_id
+
+                if all(report in self for report in tag_reports):
```
[the remark this answers is not in the document]
Comment (oco-odoo) [reply addons/account/models/account_report.py:159]: The comment explains it ;) 

It's not a matter of number of reports. Simply put: if we change the country of some reports, we want to also change the country of the related tags if they do not belong to another report. Else, we create new tags for the new country.

File: odoo/addons/base/models/ir_actions.py [hunk 12]
```diff
@@ -148,11 +148,17 @@ def _for_xml_id(self, full_xml_id):
...
+        """ Returns the action content for the provided action record.
+        """
+        self.ensure_one()
+        readable_fields = self._get_readable_fields()
         return {
             field: value
-            for field, value in action.items()
-            if field in record._get_readable_fields()
+            for field, value in self.sudo().read()[0].items()
+            if field in readable_fields
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_actions.py:161]: This doesn't seem to be used anywhere, either here or in the enterprise PR. What's the purpose, is it just to call `record._get_readable_fields()` only once? In that case, why not just extract that to a local and keep the rest untouched?

Comment (Levizar) [reply odoo/addons/base/models/ir_actions.py:161]: It seems to be used here:
https://github.com/odoo-dev/enterprise/blob/718e74f4077e28b9546018e16a53147dc4d20d28/account_reports/models/account_report.py#L2940

File: addons/account/models/account_report.py [hunk 13]
```diff
@@ -0,0 +1,643 @@
...
+    ('monetary', "Monetary"),
+    ('percentage', "Percentage"),
+    ('integer', "Integer"),
+    ('float', "Float"),
+    ('date', "Date"),
+    ('datetime', "Datetime"),
+    ('none', "No Formatting"),
+]
+
+DOMAIN_REGEX = re.compile(r'(-*sum)\((.*)\)')
```
Comment (xmo-odoo) [inline addons/account/models/account_report.py:20]: Why `-*`? Is `-----sum(...)` really somethiong we care for or about?

File: addons/account/models/account_report.py [hunk 14]
```diff
@@ -0,0 +1,643 @@
...
+    def copy(self, default=None):
+        '''Copy the whole financial report hierarchy by duplicating each line recursively.
+
+        :param default: Default values.
+        :return: The copied account.report record.
+        '''
+        self.ensure_one()
+        if default is None:
+            default = {}
+        default.update({'name': self._get_copied_name()})
```
Comment (xmo-odoo) [inline addons/account/models/account_report.py:186]:
```suggestion
        default['name'] = self._get_copied_name()
```

File: addons/account/models/account_report.py [hunk 15]
```diff
@@ -0,0 +1,643 @@
...
+                raise ValidationError(_("Groupby feature isn't supported by aggregation engine."))
+
+    def _copy_hierarchy(self, code_mapping, report=None, copied_report=None, parent=None):
+        ''' Copy the whole hierarchy from this line by copying each line children recursively and adapting the
+        formulas with the new copied codes.
+
+        :param report: The financial report that triggered the duplicate.
+        :param copied_report: The copy of the report.
+        :param parent: The parent line in the hierarchy (a copy of the original parent line).
+        :param code_mapping: A dictionary keeping track of mapping old_code -> new_code
```
Comment (xmo-odoo) [inline addons/account/models/account_report.py:290]: That parameter is a bit difficult to grok at first glance, and possibly dangerous. Maybe it should be hidden in a "copy state" object so nobody is tempted to pass in external data? From what I understand its role is simply to memo the copy state / previous renamings between line copies, that correct?

**Review by xmo-odoo — COMMENTED** [review]
A few minor notes, though I find the `code_mapping` parameter unclear at first glance, and possibly easy to misuse. Might be worth thinking about a different name, and possibly wrapping it in an opaque-ish object which ensures this is always initialised to an empty dict.

File: addons/account/models/account_report.py (hunk 15 above)
Comment (smetl) [reply addons/account/models/account_report.py:293]: It's simply the mapping from the old report line code to the new line from my understanding

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**